File: macos/TaskStrips/Sources/TaskStrips/Sync/BoardSyncService.swift
import Foundation
import SwiftData

/// One sync, from the outside: pick a transport, run it, remember what happened.
///
/// Mirrors BoardSyncService.kt, with the one difference the user chose: this device adopts on its
/// first sync. The two boards already overlap and were never given matching ids, so merging them
/// would union them into duplicates of everything on both. The phone wins; this is the side that
/// gives way, once.
@MainActor
enum BoardSyncService {

    private static let hasSyncedKey = "boardSync.hasSynced"
    private static let lastSummaryKey = "boardSync.lastSummary"

    static var hasSyncedBefore: Bool {
        UserDefaults.standard.bool(forKey: hasSyncedKey)
    }

    static var lastSummary: String? {
        UserDefaults.standard.string(forKey: lastSummaryKey)
    }

    /// Runs against whichever route the synced notes are already set to use, so there is one
    /// answer to "where does this machine sync" rather than two that can disagree.
    static func run(context: ModelContext, passphrase: String? = nil) async -> BoardSyncOutcome {
        var scoped: URL?
        defer { if let scoped { SyncFolder.endAccess(scoped) } }

        let transport: BoardTransport
        switch SyncFolder.mode {
        case .folder:
            guard let folder = SyncFolder.resolve() else {
                return BoardSyncOutcome(failure: "Pick a folder to sync through.")
            }
            scoped = folder
            transport = FolderBoardTransport(folder: folder)
        case .drive:
            guard let client = try? await DriveSession.shared.client() else {
                return BoardSyncOutcome(failure: "Sign in from the board's Drive window first.")
            }
            transport = DriveBoardTransport(client: client)
        }

        do {
            let outcome = try await BoardSyncRunner(
                context: context,
                transport: transport,
                passphrase: passphrase
            ).run(hasSyncedBefore: hasSyncedBefore, adoptsOnFirstSync: true)

            try context.save()
            // Only a sync that got somewhere counts as one. Marking a failure would tell the
            // first-sync rule this machine has synced when it hasn't — and that rule is the only
            // thing standing between an unreachable folder and a board replaced by nothing.
            UserDefaults.standard.set(true, forKey: hasSyncedKey)
            UserDefaults.standard.set(outcome.summary, forKey: lastSummaryKey)
            return outcome
        } catch {
            return BoardSyncOutcome(failure: error.localizedDescription)
        }
    }
}

extension BoardSyncOutcome {
    /// Set when the sync couldn't run at all, as opposed to running and finding nothing to do.
    var failure: String? {
        get { failureReason }
        set { failureReason = newValue }
    }

    init(failure: String) {
        self.init()
        self.failureReason = failure
    }
}

File: app/src/main/java/com/saud/taskstrip/sync/BoardSyncService.kt
package com.saud.taskstrip.sync

import android.content.Context
import com.saud.taskstrip.backup.BackupPassphraseStore
import com.saud.taskstrip.backup.DriveAuthHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * One sync, from the outside: get a token, run it, remember what happened.
 *
 * The phone never adopts. Whichever device gives way on a first sync is the user's choice, and it
 * is the Mac — so this always merges, and the only thing hasSyncedBefore does here is stay true.
 */
object BoardSyncService {

    suspend fun run(context: Context): BoardSyncOutcome = withContext(Dispatchers.IO) {
        val account = DriveAuthHelper.signedInAccount(context)
            ?: return@withContext BoardSyncOutcome(failure = "Sign in to Google Drive first.")
        val token = DriveAuthHelper.getAccessToken(context, account)
            ?: return@withContext BoardSyncOutcome(failure = "Couldn't get a Drive token.")

        val outcome = runCatching {
            BoardSyncRunner(
                context = context,
                transport = DriveBoardTransport(token),
                // The same passphrase the backup uses. Without one a credential's password stays
                // on this device rather than travelling in the clear — see BoardSyncRunner.
                passphrase = BackupPassphraseStore.get(context)
            ).run(
                hasSyncedBefore = BoardSyncPrefs.hasSyncedBefore(context),
                adoptsOnFirstSync = false
            )
        }.getOrElse { error ->
            BoardSyncOutcome(failure = error.message ?: "Sync failed.")
        }

        // Recorded only when the sync got somewhere *and* there was a board on the other side.
        // The phone never adopts, so this changes nothing here — but the two services answer the
        // same question and should answer it the same way. See BoardSyncService.swift, where
        // getting this wrong cost the Mac its one chance to adopt.
        if (outcome.failure == null && outcome.metRemoteBoard) {
            BoardSyncPrefs.markSynced(context, outcome.summary)
        }
        outcome
    }
}
